kubeadm/v1.15.0/KubeCluster.ps1

```diff
@@ -0,0 +1,273 @@
+        $bin -join -ConfigFile kubecluster.json                              Joins the Windows node to existing cluster
+        $bin -reset -ConfigFile kubecluster.json                             Resets and cleans up this Windows worker node
+    "
+
+    Write-Host $usage
+}
+
+function ReadKubeclusterConfig($ConfigFile)
+{
+    # Read the configuration and initialize default values if not found
+    $Global:ClusterConfiguration = ConvertFrom-Json ((GetFileContent $ConfigFile -ErrorAction Stop) | out-string)
+    if (!$Global:ClusterConfiguration.Install)
+    {
+        $Global:ClusterConfiguration | Add-Member -MemberType NoteProperty -Name Install -Value @{ 
+            Destination = "$env:ALLUSERSPROFILE\Kubernetes";
+        }
+    }
+
+    if (!$Global:ClusterConfiguration.Kubernetes)
+    {
+        throw "Kubernetes information missing in the configuration file"
+    }
+    if (!$Global:ClusterConfiguration.Kubernetes.Source)
+    {
+        $Global:ClusterConfiguration.Kubernetes | Add-Member -MemberType NoteProperty -Name Source -Value @{
+            Release = "1.15.0";
+        }
+    }
+    if (!$Global:ClusterConfiguration.Kubernetes.ControlPlane)
+    {
+        throw "Control-plane node information missing in the configuration file"
+    }
+
+    if (!$Global:ClusterConfiguration.Kubernetes.Network)
+    {
+        $Global:ClusterConfiguration.Kubernetes | Add-Member -MemberType NoteProperty -Name Network -Value @{
+            ServiceCidr = "10.96.0.0/12";
+            ClusterCidr = "10.244.0.0/16";
+        }
+    }
+
+    if (!$Global:ClusterConfiguration.Cni)
+    {
+        $Global:ClusterConfiguration | Add-Member -MemberType NoteProperty -Name Cni -Value @{
+            Name = "flannel";
+            Plugin = @{
+                Name = "vxlan";
+            };
+            InterfaceName = "Ethernet";
+        }
+    }
+
+    if ($Global:ClusterConfiguration.Cni.Plugin.Name -eq "vxlan")
+    {
+        if (!$Global:ClusterConfiguration.Kubernetes.KubeProxy)
+        {
+            $Global:ClusterConfiguration.Kubernetes | Add-Member -MemberType NoteProperty -Name KubeProxy -Value @{
+                    Gates = "WinOverlay=true";
+            }
+        }
+    }
+
+    if (!$Global:ClusterConfiguration.Cri)
+    {
+        $Global:ClusterConfiguration | Add-Member -MemberType NoteProperty -Name Cri -Value @{
+            Name = "dockerd";
+            Images = @{
+                Pause = "mcr.microsoft.com/k8s/core/pause:1.0.0";
+                Nanoserver = "mcr.microsoft.com/windows/nanoserver:1809";
+                ServerCore = "mcr.microsoft.com/windows/servercore:ltsc2019";
+            }
+        }
+    }
+}
+function LoadPsm1($Path)
+{
+    $tmpPath = [io.Path]::Combine([System.IO.Path]::GetTempPath(), [io.path]::GetFileName($Path))
+    Invoke-WebRequest $Path -o $tmpPath
+    Import-Module $tmpPath  -DisableNameChecking
```

Comment:
can you check if this already exists before downloading it? When testing PRs like this it's easier if you can pull all the files down and test together.

```
if (!(Test-Path $tmpPath)) { invoke-webrequest... }
```

Reply:
In the case of simply checking if the file exists and skipping pulling it down we would need to do a version check on the file incase the one on disk is older than the one upstream. I would suggest always pull down for simplicity. And yes pulling everything at the start makes more sense than doing them one by one. Ideally pull a tar ball or ZIP file and unpack it locally, much more efficient IMHO

Reply:
Switched to using DownloadFile which will check if it already exists. Per offline discussions with Patrick and m2, we're continuing with using Test-Path since it'll allow for developers to more easily use their custom files